ledger/complete/compactor.go

```diff
@@ -0,0 +1,393 @@
+	}
+	checkpoints, err := checkpointer.Checkpoints()
+	if err != nil {
+		return fmt.Errorf("cannot list checkpoints: %w", err)
+	}
+	if len(checkpoints) > int(checkpointsToKeep) {
+		// if condition guarantees this never fails
+		checkpointsToRemove := checkpoints[:len(checkpoints)-int(checkpointsToKeep)]
+
+		for _, checkpoint := range checkpointsToRemove {
+			err := checkpointer.RemoveCheckpoint(checkpoint)
+			if err != nil {
+				return fmt.Errorf("cannot remove checkpoint %d: %w", checkpoint, err)
+			}
+		}
+	}
+	return nil
+}
+
+// processTrieUpdate writes trie update to WAL, updates activeSegmentNum,
+// and gets tries from checkpointQueue for checkpointing if needed.
+// It also sends WAL update result and waits for trie update completion.
+func (c *Compactor) processTrieUpdate(
+	update *WALTrieUpdate,
+	checkpointQueue *CheckpointQueue,
+	activeSegmentNum int,
+	nextCheckpointNum int,
+) (
+	_activeSegmentNum int,
+	checkpointNum int,
+	checkpointTries []*trie.MTrie,
+) {
+
+	// RecordUpdate returns the segment number the record was written to.
+	// Returned segment number (>= 0) can be
+	// - the same as previous segment number (same segment), or
+	// - incremented by 1 from previous segment number (new segment)
+	segmentNum, skipped, updateErr := c.wal.RecordUpdate(update.Update)
+
+	// This ensures that updated trie matches WAL update.
+	defer func(updateResult error) {
+		// Send result of WAL update
+		update.ResultCh <- updateResult
+
+		// Wait for updated trie
+		trie := <-update.TrieCh
+		if trie == nil {
+			c.logger.Error().Msg("compactor failed to get updated trie")
+			return
+		}
+
+		checkpointQueue.Push(trie)
+	}(updateErr)
+
+	if activeSegmentNum == -1 {
+		// Recover from failure to get active segment number at initialization.
+		return segmentNum, -1, nil
+	}
+
+	if updateErr != nil || skipped || segmentNum == activeSegmentNum {
+		return activeSegmentNum, -1, nil
+	}
+
+	// In the remaining code: segmentNum > activeSegmentNum
+
+	// active segment is finalized.
+
+	// Check new segment number is incremented by 1
+	if segmentNum != activeSegmentNum+1 {
+		c.logger.Error().Msg(fmt.Sprintf("compactor got unexpected new segment numer %d, want %d", segmentNum, activeSegmentNum+1))
+	}
+
+	// Update activeSegmentNum
+	prevSegmentNum := activeSegmentNum
+	activeSegmentNum = segmentNum
+
+	if nextCheckpointNum > prevSegmentNum {
+		// Not enough segments for checkpointing
+		return activeSegmentNum, -1, nil
+	}
+
+	// In the remaining code: nextCheckpointNum == prevSegmentNum
+
+	// Enough segments are created for checkpointing
+
+	// Get tries from checkpoint queue.
+	// At this point, checkpoint queue contains tries up to
+	// last update (logged as last record in finalized segment)
+	// It doesn't include trie for this update
+	// until updated trie is received and added to checkpointQueue.
+	tries := checkpointQueue.Tries()
+
+	checkpointNum = nextCheckpointNum
+
+	return activeSegmentNum, checkpointNum, tries
```

Comment:
Does this function need to return `checkpointNum`? It seems if it ever returns, it's always equal to the `nextCheckpointNum` from the input. The call could just use the input argument, instead of the output of this function

Reply:
> Does this function need to return checkpointNum?

Not exactly, but I return it anyway as `checkpointNum` along with `tries` to checkpoint because the variable is called `nextCheckpointNum` in the caller and it can be confusing.  Before doing this, I thought about renaming `nextCheckpointNum` in the caller but chose this way because it seemed easier to read this way.